Blog: 5 ways to improve your analytics logging

Lior539 opened the pull request:
This blog is based off a conversation I had with Anna Debenham and [her docs](https://posthog.slack.com/archives/C01FHN8DNN6/p1692800931745099?thread_ts=1692615904.178599&cid=C01FHN8DNN6) on how to build a good data foundation. Since its mostly based off her docs, I opted to ghost-write it and credit her as the author.

Looking for feedback on:
- I want to make sure we're giving Anna enough credit. Although it shows as the author of the post, with the new redesign of the blog page, the author is nowhere to be seen. Thus I've added a callout in the beginning, so would like feedback on that 
- @joethreepwood, I've added Boldstart in the callout too. Let me know if we want to coordinate with them on this too somehow?
- @stuti90, is there anything else you think is worth including?

Lior539 commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+---
+date: 2023-09-06
+title: 5 ways to improve your analytics data
```
Other titles I considered:
- How to build a good product analytics foundation
- How to log analytics you can trust
- 5 quick wins to log better data

Lior539 replied:
Thanks! Will give it some thought.

joethreepwood commented:
I’m OOO for the rest of this week after today, so probably best for you to coordinate directly with @maban if you can @Lior539 -- waiting for me will just slow this down if this merges while I'm away. 

I'd suggest we run a Twitter thread and LinkedIn which tags her and Boldstart so that they can amplify it and push their own thread if they wish. Boldstart also runs a newsletter, so this may be appropriate to feature there if Anna is happy to do that. 

Defer to @andyvan-ph if this would be a good fit for Product for Engineers, but we can definitely add this to the PostHog for Startups newsletter -- we send that at the end of the month, so you can leave it to me to add in when the time comes.

joethreepwood commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+  - Product analytics
+---
+
+> This is a guest post by [Anna Debenham](https://www.linkedin.com/in/anna-debenham/). Anna is a former developer and director of product at [Snyk](https://snyk.io/), the open-source cybersecurity company. She is currently an operating partner at [Boldstart](https://boldstart.vc/), a VC firm supporting developer-first, infra, and saas startups. 
```
```suggestion
> This is a guest post by [Anna Debenham](https://www.linkedin.com/in/anna-debenham/). Anna is a former developer and director of product at [Snyk](https://snyk.io/), the open-source cybersecurity company. She is currently an operating partner at [boldstart](https://boldstart.vc/), a VC firm supporting developer-first, infra, and saas startups. 
```

I believe they prefer it styled like this

joethreepwood commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+
+A reverse proxy enables you to send events to your analytics provider using your domain instead of your provider's domain. This makes events less likely to be intercepted by tracking blockers.
+
+> **Tip:** If you're using PostHog to capture events, you can read our docs on [how to set up a reverse proxy](https://posthog.com/docs/advanced/proxy).
```
```suggestion
> **Tip:** If you're using PostHog to capture events, you can [set up a reverse proxy](https://posthog.com/docs/advanced/proxy).
```

Simpler and removes the mixed tense (can't say 'our' if Anna is the author)

maban commented:
[Opened a PR here](https://github.com/PostHog/posthog.com/pull/6675) with some additional content.

There are some areas I want to elaborate on but I'd really like your opinion on.

The *prefer backend to frontend events* could really do with some logic on when to use front-end and when to use back-end events. This came about when I was working with a team and trying to understand how many total users there were, but the team's front-end events were showing a different number than the database because the event was added later on, so it didn't have historical usage. I've also seen where it's a different number because it doesn't include people who signed up a different way (like being invited by another user).

This would be great as a list like:

Examples of when to use a front-end event:
* Tracking button clicks
* Tracking when a user has reached an activation point

Examples of when to use a back-end event:
* Tracking the total number of accounts.
* Tracking the number of members in a team within accounts.

But now I think about it, maybe it's not events, maybe it's more around when to fetch client-side analytics and when to fetch the metrics straight from your database. Does PostHog let you create insights that includes data from a database? It would be good to link to an article on that if so.

ivanagas commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+
+## 5. Filter out internal users
+
+When developing or testing your app, your actions skew your data and paint an inaccurate picture of actual user activity. Thus, it's important to filter out events sent by your own team (although this becomes less important as your user base grows since you're less likely to skew your data then).
```
```suggestion
When developing or testing your app, your actions skew your data and paint an inaccurate picture of actual user activity. Thus, it's important to filter out events sent by your own team (this becomes less important as you grows as you're less likely to skew your data then).
```

ivanagas commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+- Filter out internal emails.
+- Add a property on your events `is_employee` or `is_test_user` and filter events where these are `true`.
+- Exclude internal IPs from your analytics tracking.
+- Filter events by domain host e.g., exclude `localhost:3000`, `staging.yourapp.com`.
+
```
```suggestion
- Filter events by domain host e.g., exclude `localhost:3000`, `staging.yourapp.com`.
- Turn off tracking in development using a `localhost`, `dev`, or config check.

```

ivanagas commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+- Exclude internal IPs from your analytics tracking.
+- Filter events by domain host e.g., exclude `localhost:3000`, `staging.yourapp.com`.
+
+> **Tip:** For a detailed guide on how to do this, see our tutorial on [how to filter out internal users in PostHog](https://posthog.com/tutorials/filter-internal-users).
```
```suggestion
> **Tip:** For a detailed guide on how to do this, see our tutorial on [how to filter out internal users in PostHog](/tutorials/filter-internal-users).
```

ivanagas commented on contents/tutorials/how-to-build-a-good-analytics-foundation.md:
empty file?

ivanagas reviewed (commented):
Looks good, some changes 😄

Lior539 commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+
+- Use present-tense verbs, e.g., "submit" and "create" instead of "submitted" and "created".
+
+- Use snake case, i.e., `signup_button_click`
```
That was my same thought actually. I looked into the events json, and this is what we do, we just format it on the frontend to look differently
e.g., In the UI we show `Feature Flag Called`, but the name of the event is `$feature_flag_called`
Same for "Capture Metrics", it's `$capture_metrics`.
Annoyingly, `Group Identify` is `$groupidentify`

ivanagas replied:
https://github.com/PostHog/posthog.com/issues/6681 👍

Lior539 commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+
+- Only use lowercase letters.
+
+- Use present-tense verbs, e.g., "submit" and "create" instead of "submitted" and "created".
```
Not sure I follow. Why do you find it awkward?

andyvan-ph replied:
I think, given this Anna's guide rather our own, we should stick with her intent. I kind of get where Ian is coming from, but the most important thing is to be consistent one way or the other.

Lior539 commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+
+In this post, we'll cover the best practices on how to build a scalable analytics foundation that you can trust (and that your data scientist doesn't hate!).
+
+## 1. Implement a naming convention
```
I've changed the wording a bit, now its under a heading "suggested naming guide"

Lior539 commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+  - After the descriptor, add a noun that explains the specific aspect of that descriptor. For instance: `user_id`, `item_price`, or `member_count`.
+  - Use `is/has` prefixes for boolean properties. For example: `is_subscribed`, `has_seen_upsell`.
+  - If the property's value is a date or timestamp, include _date or _timestamp at the end. For example, `user_creation_date`, `last_login_timestamp`.
+
```
I think its okay to exclude. We have 2 callouts already the promote PostHog, so I dont want to go overboard

Lior539 commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,102 @@
+
+Thus, when possible, you should log events on your server instead of on your client.
+
+That doesn't mean that frontend analytics are entirely useless, though. As long as you're aware that they're only showing a segment of your data, they can still be a useful guide. Just make sure you're validating any business critical metrics against backend events or what's stored in your database! For example, instead of relying on frontend events to see how many users have signed up, rather qu ...
```
reworked this section

Lior539 commented:
Thank you @maban for your PR and feedback! I've incorporated the changes from your PR into this.

> The prefer backend to frontend events could really do with some logic on when to use front-end and when to use back-end events.

Good idea! I've added this now.

> Does PostHog let you create insights that includes data from a database? It would be good to link to an article on that if so.

Unfortunately not :/

maban commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,132 @@
+## 5. Version your events
+
+As your application evolves, so do the events you track. Implementing a versioning system for events ensures you can easily distinguish between older and newer events, especially when significant changes occur.
+
```
Not in love with this heading, but it felt better than "Summary".

maban commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,132 @@
+
+> This is a guest post by [Anna Debenham](https://www.linkedin.com/in/anna-debenham/). Anna is a former developer and director of product at [Snyk](https://snyk.io/), the open-source cybersecurity company. She is currently an operating partner at [boldstart](https://boldstart.vc/), a VC firm supporting developer-first, infra, and saas startups. 
+
+Any data scientist will tell you the worst part of their job is dealing with messy and unstructured analytics data. Usually, this is caused by software engineers carelessly capturing analytics – without a second thought about how the data will be used. The result is bad data that leads to bad decisions.
```
Softened this sentence a bit as it felt a bit finger-pointy.

maban commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,132 @@
+
+Below is a guide to help you decide when to rely on frontend or backend analytics:
+
+- Use frontend analytics if you want to:
```
Softened this. I didn't want it to sound like frontend events are inferior, when a lot of your product is based on it!

maban commented:
I made some tweaks… I think? Seems I've completely forgotten how to be an engineer and I thought I'd try this new codespaces thing and it's making me feel old.

![A dog that has no idea what it's doing.](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExNWJrZ3N0bGViN3F6dmRvNXlscDFmaTZuZjNoeHJsM3piODVpNjMwYiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/xDQ3Oql1BN54c/giphy-downsized.gif)

maban commented:
Here's my PR with proposed additions/tweaks: https://github.com/PostHog/posthog.com/pull/6679
Mostly softening the language or elaborating on bits.

stuti90 commented:
@Lior539 love this blog post! can't wait for it to go live 🚀 No suggestions from me - think you've covered everything!

andyvan-ph commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,133 @@
+---
+date: 2023-09-06
```
```suggestion
date: 2023-09-12
```

andyvan-ph commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,133 @@
+---
+date: 2023-09-06
+title: 5 ways to improve your analytics data
```
```suggestion
title: How to structure your analytics data so it's usable
```

I feel like this is the most "what it says on the tin" title for this. Talk of "foundations" is too abstract.

andyvan-ph commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,133 @@
+  - anna-debenham
+featuredImage: ../images/blog/super-hog-pink.png
+featuredImageType: full
+category: General
```
```suggestion
category: Engineering
```

andyvan-ph commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,133 @@
+
+A reverse proxy enables you to send events to your analytics provider through your domain instead of your provider's. This makes events less likely to be intercepted by tracking blockers.
+
+> **Tip:** If you're using PostHog to capture events, here's how you can [set up a reverse proxy](/docs/advanced/proxy).
```
```suggestion
> 💡 **PostHog Tip:** If you're using PostHog to capture events, here's how you can [set up a reverse proxy](/docs/advanced/proxy).
```

andyvan-ph commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,133 @@
+> 
+> For a detailed guide on how to do this in PostHog, see the tutorial on [how to filter out internal users in PostHog](/tutorials/filter-internal-users).
+
+## 3. Prefer backend to frontend tracking
```
Should we add something in this section about:

1. The backend libraries we support / setup guides for backend tracking in PostHog
2. How to label backend events vs frontend events in PostHog? 

On the latter, I assume that's a use case Data Management and applying tags to events to show where they were captured?

Lior539 replied:
On both points: We have 2 callouts already (out of 5 points), so I'm reluctant to add more posthog specific information, in order to avoid pushing posthog too hard in this post

> How to label backend events vs frontend events in PostHog?

There are a few ways to do this in PostHog e.g. posthog logs a property from which library the event came from

andyvan-ph reviewed (approved):
A few suggestions, but this looks good.

Lior539 commented on contents/blog/5-ways-to-improve-analytics-data.md:
```diff
@@ -0,0 +1,133 @@
+---
+date: 2023-09-13
+title: 5 ways to improve your analytics data
```
Thanks both for the suggestions. I went back and forth on this a bunch of times. In the end I think "5 ways to improve your product analytics data" is the most eye catching
```suggestion
title: 5 ways to improve your product analytics data
```
